### ISSUE TITLE
Schema enhancements "Frameworks"

### ISSUE THREAD

**Issue by tmfrnz** [impactoss/impactoss-server#326]
The DB schema needs to be enhanced to support frameworks which can be understood as different types of objectives, which will allow merging the SDG targets into the recommendations table and also include other types of recommendations (ie national implementation targets). The recommendations table will now store not only recommendations in the strict sense but any type of targets or objectives (that are all all aspirational). Later we could consider changing the name but suggest to keep for now.

The schema changes, endpoints and permissions are documented here
https://github.com/nmrf/sadata/issues/41#issuecomment-705146834

The tasks are 
- [ ] update db **schema** as specified above, creating new and updating existing tables, as well as including indices on foreign keys
- [ ] create **models** for the new tables
- [ ] set up and update **tests** for new and updated tables/models
- [ ] set up **serializers** for the new tables/models, also add new fields/remove obsolete fields for updated tables/models
- [ ] set up **controllers** for new tables/models. 
- [ ] set up and update **policies** for new and updated tables/models

Again see the https://github.com/nmrf/sadata/issues/41#issuecomment-705146834 for endpoints

In summary the new tables frameworks and the join tables framework_frameworks and taxonomy_frameworks are only to be changed during set up and not by the front-end. Therefore the only endpoint required is READ. They are similar to the existing taxonomies table.

All other new tables are simple join tables to allow many-to-many relationships between different concepts. For the join tables best use the existing sdgtarget_recommendations table/model as an example

**Comment by oventi** [comment]
# tables
- [ ] CREATE table frameworks

# endpoints/permissions
- [ ] frameworks, read: public (no authentication or role required for reading the frameworks)

**Comment by oventi** [comment]
@tmfrnz  the tables `frameworks_frameworks` and `recommendations_recommendations` are not needed because the relationships can be achieved in the same table.

**Comment by oventi** [comment]
@tmfrnz I found more complexity regarding the indicators and recommendations relationship. They seem to be tied via the measures table. I will remove that and tie them directly, but that will probably require extra testing.

**Comment by tmfrnz** [comment]
> @tmfrnz the tables `frameworks_frameworks` and `recommendations_recommendations` are not needed because the relationships can be achieved in the same table.

@oventi well we want to enable many-to-many relationships so not sure how you would want to achieve this?

**Comment by tmfrnz** [comment]
> @tmfrnz I found more complexity regarding the indicators and recommendations relationship. They seem to be tied via the measures table. I will remove that and tie them directly, but that will probably require extra testing.

@oventi could the recommendations have both, direct indicators and indicators through measures?

This relationship is used only for generating email reminders for "category managers" (manager of a category tagging a recommendation linked to an indicator (through a measure) will get an email if a scheduled progress report is missing for that indicator).

**Comment by oventi** [comment]
> > @tmfrnz I found more complexity regarding the indicators and recommendations relationship. They seem to be tied via the measures table. I will remove that and tie them directly, but that will probably require extra testing.
> 
> @oventi could the recommendations have both, direct indicators and indicators through measures?
> 
> This relationship is used only for generating email reminders for "category managers" (manager of a category tagging a recommendation linked to an indicator (through a measure) will get an email if a scheduled progress report is missing for that indicator).
> 
> Here the category indicator link:
> 
> https://github.com/impactoss/impactoss-server/blob/a9227873092b68e78b5ce10bd95024723a826b83/app/models/category.rb#L11

OK, so I will revert the removed association and add another one. From what I read, Rails supports these multiple associations.

**Comment by oventi** [comment]
@tmfrnz on Indicators, recommendations via measures will be available via `recommendations` property, and recommendations via Indicators Recommendations (the newly created joined table) will be accessed via `direct_recommendations`

**Comment by oventi** [comment]
@tmfrnz to create and delete


```
POST /recommendations/[id]/recommendations
{
    "other_recommendation_id": 22
}

DELETE /recommendations/[id]/recommendations/22
```

```
POST /indicators/[id]/recommendations
{
    "recommendation_id": 22
}

DELETE /indicators/[id]/recommendations/22
```

**Comment by tmfrnz** [comment]
> @tmfrnz to create and delete
> 
> ```
> POST /recommendations/[id]/recommendations
> {
>     "other_recommendation_id": 22
> }
> 
> DELETE /recommendations/[id]/recommendations/22
> ```
> 
> ```
> POST /indicators/[id]/recommendations
> {
>     "recommendation_id": 22
> }
> 
> DELETE /indicators/[id]/recommendations/22
> ```

Hi @oventi, I am not quite sure I understand, but I think the endpoints should be consistent with the other "join table" endpoints, so
 
POST /recommendation_recommendations/
{
     "recommendation_id": 1,
     "other_recommendation_id": 22
}

and 

DELETE /recommendation_recommendations/[id]

where the id is the unique id of that relationship

**Comment by tmfrnz** [comment]
@oventi

so we want specific controllers for these relationships similar to what we have for 
https://github.com/impactoss/impactoss-server/blob/frameworks-schema/app/controllers/sdgtarget_recommendations_controller.rb

also we will need serializers for reading the relationship pairs, again similar to
https://github.com/impactoss/impactoss-server/blob/frameworks-schema/app/serializers/sdgtarget_recommendation_serializer.rb

**Comment by tmfrnz** [comment]
> ## RCD
> * Read: public (no authentication)
> * Create: user role manager or admin (default application record)
> * Delete: user role manager or admin (default application record)
> 
> * [x]  frameworks | PUBLIC
> * [ ]  recommendations_recommendations | Read
> * [ ]  recommendations_recommendations | Create
> * [ ]  recommendations_recommendations | Delete
> * [ ]  indicators_recommendations | Read
> * [ ]  indicators_recommendations | Create
> * [ ]  indicators_recommendations | Delete
> * [ ]  frameworks_taxonomies | PUBLIC
> * [ ]  frameworks_frameworks | PUBLIC

@oventi: note that I have unchecked all these endpoints as I do not think they have been implemented the right way

These actually state how the paths should look:

* /recommendations_recommendations | Read
* /recommendations_recommendations | Create
* /recommendations_recommendations | Delete
* /indicators_recommendations | Read
* /indicators_recommendations | Create
* /indicators_recommendations | Delete
* /frameworks_taxonomies | Read
* /frameworks_frameworks | Read

**Comment by tmfrnz** [comment]
I think this is the important bit that was missed

> For the join tables best use the existing sdgtarget_recommendations table/model as an example

**Comment by nrankin** [comment]
hey @tmfrnz I have done frameworks and framework_frameworks, can you take a look, particularly as the spec/requests/framework_frameworks_spec.rb  and tell me if this is what you were expecting before I go and replicate that pattern all over :)

> These actually state how the paths should look:
> ...
> ...
> * /frameworks_frameworks | Read

I have made the path framework_frameworks (for table frameworks_frameworks). I realise that you have explicitly written "frameworks_frameworks" above. However, the singular/plural pattern for the path seems consistent with the sdgtarget_recommendations path (for table sdgtargets_recommendations) and it also seems more straight forward from a rails perspective.
But I can try to change it if that isn't what you need.

**Comment by tmfrnz** [comment]
@nrankin looks pretty good generally but I have left a few comments/questions for you about consistent spelling and unfortunately also found a bunch of small inconsistencies in Andres' previous changes (see e.g. https://github.com/impactoss/impactoss-server/pull/327/commits/be0c86369e2a90032c3ddc082a7fa02e242eb7b0#r532654073)

Basically it appears to go back to the first comment above (https://github.com/impactoss/impactoss-server/issues/326#issuecomment-706969330) where the inconsistent spelling for table names was introduced when they were wrongly copied from https://github.com/nmrf/sadata/issues/41. 

Unfortunately this means that all these join models and tables are now inconsistent and will probably need to change? So gutted I did not notice this earlier!!

**Comment by tmfrnz** [comment]
To be clear, here the corrected list of join table and endpoint names:

- recommendation_recommendations
- indicator_recommendations (originally I specified recommendation_indicators but not sure why Andres might have flipped this)
- framework_taxonomies
- framework_frameworks

**Comment by tmfrnz** [comment]
@nrankin wondering how to best deal with these f*cked up migrations: remove & redo or modify?

**Comment by nrankin** [comment]
@tmfrnz I'm gutted I didn't notice too... I was deliberately looking to the sdgtarget_recommendations as an example as you said and trying to check for where the previous changes had gone wrong, clearly I missed it on the table/model names :( 

It is best to just run new migrations that change the table names, and I will remove all the `self.table_name = "recommendations_recommendations"` etc that Andrés put in as they won't be required as now the default rails pluralization will work fine. Then there will be some other changes but hopefully it's not too bad :) Good tests make is easier.

**Comment by tmfrnz** [comment]
@nrankin, it would be great if you could make these changes to the table names and all their references (and remove where they are no longer needed. Thanks!

**Comment by nrankin** [comment]
@tmfrnz no worries, I'm keen to get this all cleaned up. Probably wont have time till the weekend but unless I come across something unexpected I think I'll smash it all out then.

**Comment by tmfrnz** [comment]
@nrankin thank you so very much =)

**Comment by nrankin** [comment]
@tmfrnz I think I have fixed everything up and made sure all the endpoints are there and tested. 
I did not manage to work out how to sign in in a request spec so only the read endpoints are tested in those. Shame as they are much more end to end and show the endpoints better for a human. But I have pretty thorough controller testing I think, so it's fine. 
Let me know if this is now more like what you need or I have missed something. Please don't worry about spellings in migrations, check instead that you think the schema looks correct :)
Actually, just realised I still need to write a few model tests... need to have a last good look through but would love your feedback on the endpoints and controller, request tests so far.

**Comment by tmfrnz** [comment]
thank you @nrankin - have had a quick look at your latest commits and the schema and am happy to confirm that it all looks pretty good. The only thing I notice that the taxonomies table is missing the framework_id [FK] field (see https://github.com/nmrf/sadata/issues/41#issuecomment-705146834 - I thought Andres had implemented this already)

I am not also not sure how to sign in in an request spec - are there no examples for the other endpoints?

**Comment by nrankin** [comment]
fantastic @tmfrnz !
I will look into the taxonomy framework_id tonight.
There were no request specs before, I have created the only ones. There are plenty of controller specs, which can successfully log in with `sign_in user` for example, however that does not work in request specs.
I just wanted to add request specs as it's rails recommendation going forward to replace controller specs with request specs. 
But, I also didn't want to spend too much time trying to figure it out since the controller specs are the status quo in this app.
Happy to have another look at it if you like tho?

**Comment by nrankin** [comment]
I've added that single framework association to taxonomies and done a bit of refactoring.
One of the things I did was switched from the deprecated "should" syntax to the new "expect_to" syntax. I've only done that for the specs for the new tables, but it's easy enough for me to change all the model specs if you like - it's just a find/replace/runthespecs and everything seems to work... just didn't want to overstep the bounds of the task at hand without an ok from you :)

**Comment by tmfrnz** [comment]
Thank you so much @nrankin for adding that additional association and also for the refactoring. Please feel free to go ahead and refactor all other model specs if it is not too much additional effort. Thanks!!

**Comment by nrankin** [comment]
ok, all done I believe @tmfrnz :)

**Comment by tmfrnz** [comment]
@nrankin thanks so much - I will have a closer look over the weekend

**Comment by tmfrnz** [comment]
@nrankin thank you again for all your help. 

I had a closer look and also tested all the "Read" endpoints there. As far as I can tell everything looks pretty good but except these things that I noticed (and mostly fixed):

* new framework_id field was missing in some serializers and policies (see https://github.com/impactoss/impactoss-server/commit/223ab3e3ffee6259127a05016fc4ae6316d124b5)
* I noticed an unspecified field "parent_id" was [added to the recommendations table that should be removed](https://github.com/impactoss/impactoss-server/commit/501116a12b36b41a1ee2c5cbe8c4e9351fce056b#diff-cbf8b25d6d853acf58fa9057841f407d29ffe90649c75cf9e3f51b2d6e3aa1d3R182). Would you be able to add the required migration to remove this? (I already [removed it from the recommendations serializer](https://github.com/impactoss/impactoss-server/commit/5f620cc6001961bf5bc825cc927b8b1b89b460de)). I think it may go back to this misunderstanding: https://github.com/impactoss/impactoss-server/issues/326#issuecomment-708126729. It looks like it has since been [dropped from the frameworks table](https://github.com/impactoss/impactoss-server/commit/6d29bf1a8a82d21f4ec9c094cdb6fa0bb76181ad#diff-cbf8b25d6d853acf58fa9057841f407d29ffe90649c75cf9e3f51b2d6e3aa1d3L64) but not the recommendations table.  

Next week I should be able to also test all the other endpoints when implementing the new schema on the client...
